### ISSUE TITLE
HTMLImport polyfill is extremely slow in IE11

### ISSUE THREAD

**Issue by eugef** [webcomponents/webcomponentsjs#597]
I've noticed, that in IE11 HTMLImport polyfill increases a delay before `WebComponentsReady` event is fired.
Even if there is no `<link rel='import'>` tag on the page.

In my case if i remove HTMLImport code from `webcomponents-lite.js` then delay drops from ~5s to ~1s.

I've also noticed that HTMLImport polyfill is very sensitive to the amount of other html/ajax request on the page. For example, if page has google adsense or any banners network, then delay can be up to ~10-20s!

So far my solution is to create cutom build of webcomponents polyfill without HTMLImport.
But maybe there is more elegant solution to disable HTMLImport polyfill?

**Comment by ebidel** [comment]
Nothing should be taking that long. I suspect there's something else going on. Can you post a jsbin that repros the issue?

**Comment by eugef** [comment]
Here is a simple example to reproduce the issue https://jsfiddle.net/wxncd3f1/

If you run it in IE11 you will see that `WebComponentsReady` event is fired only after all images are loaded. If you open Network tab in Developers tolls it would be clear that event is fired after all pending http requests are resolved.
On my machine with 60 images I see delay ~ 8seconds.

And there is direct correlation between amount of http request and delay before `WebComponentsReady` event.

For Firefox it is not the case, `WebComponentsReady` event is fired immediately after `DomContentReady`.

**Comment by eugef** [comment]
@ebidel and here is example with custom build of `webcomponents-lite.js` without polyfill for `HTMLImport` - https://jsfiddle.net/y5v7p3n0/3/

As you can see `WebComponentsReady` event is fired almost immediately (~80 ms on my machine).

**Comment by ebidel** [comment]
I'm no expert in the poyfills, but I _believe_ this is due to your demo setup.

The difference between native HTML imports and polyfilling them is that the polyfill needs to script! It sets up a MutationObserver to detect imports on the page and waits for the [document to be "complete"](https://github.com/webcomponents/webcomponentsjs/blob/master/src/HTMLImports/base.js#L88) before doing its thing.

In your demo, you're running a bunch of script at the bottom of the page that appends images to the DOM on page load. Scripts block parsing. This in turn, delays the polyfill from firing the `HTMLImportsLoaded` and `WebComponentsReady`. It also artificially delays `DOMContentLoaded` and `load` events!

For example, in Edge, using 100 images delays all the events:

<img width="1223" alt="screen shot 2016-08-19 at 11 56 15 am" src="https://cloud.githubusercontent.com/assets/238208/17821077/98096d8a-6604-11e6-89f7-6e22581e9084.png">

If delay the loading of those images (e.g. wrap all your code in an `setTimeout(function() { ... }, 1000);`, you'll see a big different and not delay all of the load events :)

**Comment by eugef** [comment]
Hi @ebidel, you just confirmed my assumption that the reason for delay is exactly the waiting for all sub resource http request to be resolved, because only after that `document.readyState` would be `complete`.

Unfortunately on the real website it is not possible to delay all images, scripts, etc to be loaded after some timeout. For example, on real website image tags are rendered server-side, not frontend-side.
Here is updated example that shows it: https://jsfiddle.net/wxncd3f1/1/

Maybe the solution would be to detect imports earlier, not waiting when  `document.readyState` is `complete`, but when it is `interactive`?
In this example it is clear that `interactive` event occurs much earlier then `complete` https://jsfiddle.net/wxncd3f1/3/
Is there a good reason why only for IE polyfill is wating for the `complete`?

**Comment by eugef** [comment]
BTW, using `HTMLImportsLoaded` instead of `WebComponentsReady` won't help, as they are fired almost  simultaneously, here you can see it  https://jsfiddle.net/wxncd3f1/4/

**Comment by ebidel** [comment]
SSR images are fine and that's what you should be doing. In the case of https://jsfiddle.net/wxncd3f1/1/, there's no blocking scripts so WCR fires at 20-25ms for me on IE11 and Edge in a VM.

**Comment by Garbee** [comment]
One trick to lazily-load the images is to not have a `src` by default and use JS post-load to add the src attribute. For example:

```html
<img data-src="https://example.com/image.jpg">
<img data-src="https://example.org/image.jpg">
<!-- Post load complete -->
<script>
  document.querySelectorAll('img[data-src]').forEach(node => {
    node.src = node.dataset.src;
  });
</script>
```

**Comment by eugef** [comment]
@ebidel if you clean the cache then SSR images would also give you delay.

@ebidel , @Garbee I think you are trying to fix the issue in the wrong way. The problem is that `WebComponentsReady` event is fired in IE only after all sub-resource http requests are resolved. So any slow http requests can prevent webcomponent to be rendered. 
And this problem doesn't exist in FF or Safari!

For existing website it is not feasible to change all http request (not only for images, but also for scripts, ajax requests, etc) to be executed after HTMLImport polyfill is executed. So now it is almost impossible to use webcomponents polyfill in IE for the websites that load lots of resources.

So far i see two possible solutions:

1. Preferred. Use for IE the same event as for other browsers  - `document.readyState === 'interactive'` instead of `document.readyState === 'complete'` [here](https://github.com/webcomponents/webcomponentsjs/blob/master/src/HTMLImports/base.js#L88)

2. Workaround. Create a new build of webcomponents polyfill without `HTMLImport`,  then it doesn't affect load time in IE  (here is an example https://jsfiddle.net/y5v7p3n0/3/, compare with https://jsfiddle.net/y5v7p3n0) 

I am going to implement the first approach and share the results.

**Comment by eugef** [comment]
@ebidel, here is demo for the fix i've proposed above https://jsfiddle.net/y5v7p3n0/4/. 
For all browsers html imports are executed when `document.readyState === 'interactive'`

I've also tested this fix with the Polymer components and didn't notice any bugs.
But of course it makes sense to run more extensive testing.

Do you know - maybe there was some reason not to use `document.readyState === 'interactive'` for IE?

**Comment by ebidel** [comment]
Not sure. @kevinpschaaf @sorvell @azakusm @garlicnation  may know the timing details.

**Comment by eugef** [comment]
I've look through the code and found [this comment](https://github.com/webcomponents/webcomponentsjs/blob/master/src/CustomElements/boot.js#L112) about IE and `document.readyState === 'interactive'`

> // When loading at readyState interactive time, bootstrap only if HTMLImports
// are not pending. Also avoid IE as the semantics of this state are unreliable.

Maybe there is some information why readyState interactive is unreliable in IE?
Maybe it is not the case for IE11?

**Comment by ebidel** [comment]
Probably related to:
- https://social.msdn.microsoft.com/forums/ie/en-us/6981866f-b548-42e4-a259-a66a5cafb380/ie11-onreadystatechange-event-not-firing
- https://connect.microsoft.com/IE/feedback/details/792880/document-readystat
- http://stackoverflow.com/questions/26834845/in-ie11-event-is-not-firing-in-javascript

Looks like it could be an issue with older versions of IE that need to be supported.

**Comment by eugef** [comment]
Hi @ebidel, looks like 1st and 3rd links are related only to iframe issue and not relevant for HTMLImport. Please correct if I am wrong, but webcomponents polyfill doesn't use iframes.

The issue described in 2nd link sounds to be resolved in IE11 and from my experience this is correct.
Though, i didn't try in IE10.

If polyfill need to support IE10 and this makes IE11 run slower - i would do more acccurate browser detection and for IE11 use `document.readyState === 'interactive'`.

What do you think?

**Comment by ebidel** [comment]
Yes, I think that'd be a good idea. I'm not sure why the polyfill was authored the way it is, but we should make IE11 faster if this issue has been fixed.

...would be nice if someone from the core team could chime in.

**Comment by eugef** [comment]
@kevinpschaaf @sorvell @azakus @garlicnation - any chances that this issue will be tackled in v1 version?

**Comment by eugef** [comment]
@kevinpschaaf @sorvell @azakus @garlicnation as i see you guys are busy with webcomponents polyfill v1, does it make sense to fix this issue for the new version (https://github.com/webcomponents/html-imports) ?

If yes - I can help with testing and improving HTMLImport polyfill for IE.

**Comment by notwaldorf** [comment]
cc-ing @valdrinkoshi as an FYI, as he's working on the `v1` polyfill

**Comment by nicolasmoise** [comment]
What is the status of this issue? I am still having the same problem using the `v1` poylfill

**Comment by valdrinkoshi** [comment]
re wait for `interactive`: this is already done in the [`html-imports v1` polyfill](https://github.com/webcomponents/html-imports/blob/master/src/html-imports.js#L621)

**Comment by TimvdLippe** [comment]
I am closing this issue per the above comment. If you are still encountering issues, please open an issue on https://github.com/webcomponents/html-imports for us to take a look.
